### DISCUSSION TITLE
How is milvus scalable?

### DISCUSSION THREAD

**Discussion by juanandreas** [milvus-io/milvus#34316] [Q&A and General discussion]
I shipped my team's data to a milvus database and we're seeing high memory usage that requires 512GiB memory instance on our ec2 instance. This is because milvus requires a collection to be loaded into memory. As our data grows, how is this scalable? Is milvus built for large datasets of this sort of magnitude?

**Comment by xiaofan-luan** [comment]
> I shipped my team's data to a milvus database and we're seeing high memory usage that requires 512GiB memory instance on our ec2 instance. This is because milvus requires a collection to be loaded into memory. As our data grows, how is this scalable? Is milvus built for large datasets of this sort of magnitude?

1. to scale，you have to deploy milvus with K8s. The largest milvus cluster we saw is more than 100 nodes, with 50B data.
2. You can use diskann to reduce memory cost. You need to cache data on NVMe SSD but if you do it correct you will be able to reduce memory by 10X.
3. the easiest way to scale is to use zilliz cloud https://zilliz.com/pricing. You can scale from 1-10B in a few click because we setup everything for you

**Reply by juanandreas** [reply]
100 nodes with 50B data does that mean 5GB of data? Our collection when loaded into memory takes up almost 500GB of ram. Not familiar with kubernetes but would operating over a cluster scale be a cost-effective solution for our scale?

**Reply by xiaofan-luan** [reply]
50B -> 50 billion vectors, sorry for the misunderstanding.
How do you deploy your cluster for now?

**Reply by yhmo** [reply]
50B data means 50 billion rows of data.

Memory usage mainly depends on:
- how many rows in the collection
- vector dimension
- index type

Let's say your vector dimension is 768, then each vector size is 768 * 4 bytes since each dimension value is represented by a float32 value.
There are 50 billion rows of data, 50 billion = 50 * 100M. So, the total size is 768 * 4 Bytes * 50 * 100M = 15TB.

There is a tool to estimate required resource according to data size: https://milvus.io/tools/sizing
![Screenshot from 2024-07-03 10-53-04](https://github.com/milvus-io/milvus/assets/2282099/ceba40b2-16d5-41c6-a008-599c011aefe7)

The tool will tell you that 16 core 128 GB x  229 query nodes are required for 50B 768 dim vectors.

**Reply by juanandreas** [reply]
@xiaofan-luan We are currently deploying a single ec2 instance to host our milvus collection. It's a very expensive instance since we require, as of last upload, 470GB of data., which implies about 500GB of memory. We expect our data to grow so having a single instance will not be scalable.

we have a 5.1M row table we want to ship to milvus. Each row has a 10 million dimension sparse vector. Here's an example of our data:

```
string_id_1	string_id_1a	13	1712274269	false	false	0.5075	{"323":1,"324":1,"330":1,"423":1,"608":1,"629":1,"635":1,"645":1,"647":1,"649":1,"801":1,"802":1,"803":0.5075,"250497":1,"752752":1,"1002112":1,"1253464":1,"1503655":1,"1505144":1,"1755071":1,"1755137":1,"2004589":1,"2005053":1,"2255564":1,"2506252":1,"3007880":1,"3258268":1,"3508870":1,"3511420":1,"4009575":1,"4760132":1,"5010350":1,"5010392":1,"5766570":1,"6259542":1,"6761060":1,"6761266":1,"7262579":1,"7262592":1,"7512554":1,"7512584":1,"7764003":1,"7766471":1,"8013474":1,"8270979":1}
```

2 follow up questions:

* Will having milvus cluster with kubernetes allow us to have data of our scale and continue to scale as our data grows?
* What is the fastest way to write data into our collection? We are currently using pyspark on databricks to manipulate our raw data before shipping to milvus. I think the most effective tool of choice to write data into milvus collection would be to use spark-milvus connector, but last time I checked, spark-milvus connector has yet to support sparse vector.

**Reply by juanandreas** [reply]
@yhmo We're using sparse vectors because most of the values in our vectors are 0. If we were to use normal vectors, we'd have a 10M dimension vector. Please see the data sample i provided above. I can't enter a 10M vector dimension value but with sparse vector, we should not have 10M long sparse vectors. Maybe only 20-30

**Reply by xiaofan-luan** [reply]
spark milvus will support sparse vector soon, @wayblink is working on that.
You can also work on import sparse vector, @bigsheeper just finished te code.

Yes, using K8s, you can easily scale milvus cluster by pod numbers
